Reorganize READMEs and make each package's directions clearer (#24)
foxglove/ws-protocol: python/README.md

# Foxglove WebSocket protocol in Python

The Foxglove WebSocket protocol is encoding-agnostic, as long as the desired encoding is supported by both client and server. It currently supports Protobuf and JSON messages.

## Installation

Install the `foxglove_websocket` Python package:

```

> $ pip install foxglove_websocket

Reviewer: I was surprised this works... I guess it is installable via both names, `foxglove-websocket` and  `foxglove_websocket`. I believe it is conventional for package names to use dashes, and apparently PyPI will [actually convert underscores to dashes](https://www.python.org/dev/peps/pep-0503/#normalized-names). PEP 8 also says modules (the thing you import) use underscores, while underscores in package names are discouraged: https://www.python.org/dev/peps/pep-0008/#package-and-module-names  So I think we should keep it with a dash here.

https://groups.google.com/g/comp.lang.python/c/Y5zcSR7wn7c
https://gist.github.com/gene1wood/9472a9d0dffce1a56d6e796afc6539b8
https://github.com/ros/rosdistro/issues/18116

# Foxglove WebSocket server

> This package provides an example server implementation of the [Foxglove WebSocket protocol](https://github.com/foxglove/ws-protocol), enabling [Foxglove Studio](https://github.com/foxglove/studio) to ingest arbitrary “live” streamed data.

Reviewer: Not just an example server, this package provides a library meant to be used when writing a custom server

```suggestion
This package provides a server implementation of the [Foxglove WebSocket protocol](https://github.com/foxglove/ws-protocol), enabling [Foxglove Studio](https://github.com/foxglove/studio) to ingest arbitrary “live” streamed data.
```

> The [Foxglove WebSocket protocol](https://github.com/foxglove/ws-protocol) supports any data encoding, as long as it is supported by both client and server. This package provides an example server implementation of the protocol, enabling [Foxglove Studio](https://github.com/foxglove/studio) to ingest arbitrary “live” streamed data.

Reviewer: Seems like the order of sentences is basically being swapped here. I find it more intuitive the original way: starting with "what is this package for", and then a second sentence which gives some details on the nature of the protocol. We could also consider just removing the "encoding-agnostic" sentence, since it's arguably a kind of low-level detail, which people can already learn by clicking on the link to the main docs page.

## Example server

Run a simple example server that publishes messages on a single `example_msg` topic, using JSON to encode message data and [JSON Schema](https://json-schema.org/) to describe the message layout:

```

> python -m foxglove_websocket.examples.simple_server

Reviewer: renamed to `json_server` on main, and added `protobuf_server`

```

To see data from this server, open [Foxglove Studio](https://studio.foxglove.dev?ds=foxglove-websocket&ds.url=ws://localhost:8765/) with a Foxglove WebSocket connection to `ws://localhost:8765/`:

<img width="676" alt="Foxglove Studio displaying data from the example server" src="https://user-images.githubusercontent.com/14237/145260376-ddda98c5-7ed0-4239-9ce4-10778ee8240b.png">

> ## Server template

Reviewer: Not sure this section is worth keeping anymore. It just points at a script that we already explained in the `Example server` section.

Author: Combining with previous section (in-line links to files).

Copy [this template code](https://github.com/foxglove/ws-protocol/blob/main/python/src/foxglove_websocket/examples/simple_server.py) into a file and run it (e.g. `python3 server.py`). Then, make the necessary adjustments to the file to customize this simple server to your desired specifications.

## Development

> Use [pipenv](https://github.com/pypa/pipenv) to manage developer dependencies and `virtualenv` while developing.

Reviewer: I might prefer just reverting this line; I think it's a little confusing to just say "while developing" since it's not clear whether we mean while developing in (a fork of) this repo, or developing your own server. Pipenv is something I found to be relatively easy to use for the dev setup here, but others might not want to use it
```suggestion
When developing or maintaining the foxglove-websocket package, it is recommended to use [pipenv](https://github.com/pypa/pipenv) to manage development dependencies and virtualenv.
```

                            "count": {"type": "number"},
                        },
                    }
                ),
            }
        )

        i = 0
        while True:
            i += 1
            await asyncio.sleep(0.2)
            await server.send_message(
                chan_id,
                time.time_ns(),
                json.dumps({"msg": "Hello!", "count": i}).encode("utf8"),
            )

if __name__ == "__main__":
    run_cancellable(main())
```
Copy [this template code](https://github.com/foxglove/ws-protocol/blob/main/python/src/foxglove_websocket/examples/simple_server.py) into a file and run it (e.g. `python3 server.py`). Then, make the necessary adjustments to the file to customize this simple server to your desired specifications.

## Development

When developing or maintaining the foxglove-websocket package, it is recommended to use [pipenv](https://github.com/pypa/pipenv) to manage development dependencies and virtualenv.
Use [pipenv](https://github.com/pypa/pipenv) to manage developer dependencies and `virtualenv` while developing.

- Run `pipenv install --dev` to create a virtualenv and install development dependencies.

> - Run `pipenv shell` to enter the virtualenv.

Reviewer: I'm ok with changing around the format of this section, but I don't want to lose the explanatory text on these commands... maybe:

```
- `pipenv install --dev` — create a virtualenv and install development dependencies
- `pipenv shell` — enter the virtualenv
```
?

from foxglove_websocket import run_cancellable
from foxglove_websocket.server import FoxgloveServer, FoxgloveServerListener
from foxglove_websocket.types import ChannelId

> 

Author: Linking directly to https://github.com/foxglove/ws-protocol/blob/main/python/src/foxglove_websocket/examples/simple_server.py instead, as it seems to be the same as what is here?

                time.time_ns(),
                json.dumps({"msg": "Hello!", "count": i}).encode("utf8"),
            )
if __name__ == "__main__":
    run_cancellable(main())
```

## Development

Use [pipenv](https://github.com/pypa/pipenv) to manage developer dependencies and `virtualenv` while developing.

```
pipenv install --dev
pipenv shell

> ```

Author: Want to minimize including info that isn't specific to this package / can be looked up via a link we include. Don't feel strongly about this though.
